tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,511 @@
+        {
+            if (Interlocked.CompareExchange(ref _initialized, 1, 0) == 1)
+            {
+                return;
+            }
+
+            // Try to resolve System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource
+            _diagnosticListenerType = Type.GetType("System.Diagnostics.DiagnosticListener, System.Diagnostics.DiagnosticSource");
+            if (_diagnosticListenerType is null)
+            {
+                // Because we cannot resolve the DiagnosticListener type we allow a later initialization.
+                // For this case we are going to do some retries with a back-off.
+                if (Interlocked.Decrement(ref _initializationRetries) > 0)
+                {
+                    Task.Delay(InitializationBackoffPerRetry, cancellationToken).ContinueWith(
+                        _ =>
+                    {
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
+                        Interlocked.Exchange(ref _initialized, 0);
+                        Initialize();
+                    },
+                        cancellationToken);
+                }
+
+                return;
+            }
+
+            // Initialize
+            var diagnosticSourceAssemblyName = _diagnosticListenerType.Assembly.GetName();
+            Log.Information("DiagnosticSource: {diagnosticSourceAssemblyNameFullName}", diagnosticSourceAssemblyName.FullName);
+
+            var version = diagnosticSourceAssemblyName.Version;
+
+            // Check if Version 5 or 6 is loaded (Uses ActivityListener implementation)
+            if (version?.Major is 5 or 6)
+            {
+                BindAndCreateDelegates();
+                ChangeActivityDefaultFormat();
+                CreateActivityListenerInstance();
+                return;
+            }
+
+            // Check if Version is 4.0.2 or greater (Uses DiagnosticListener implementation)
+            if (version >= new Version(4, 0, 2))
+            {
+                BindAndCreateDelegates();
+                ChangeActivityDefaultFormat();
+                CreateDiagnosticSourceListenerInstance();
+                return;
+            }
+
+            Log.Information("An activity listener was found but version {version} is not supported.", version.ToString());
+
+            void BindAndCreateDelegates()
+            {
+                // If we found a supported version we load the shared types.
+                _activityType = Type.GetType("System.Diagnostics.Activity, System.Diagnostics.DiagnosticSource");
+                if (_activityType is null)
+                {
+                    throw new NullReferenceException("Activity type cannot be found!");
+                }
+
+                _onShouldListenToMethodInfo = typeof(ActivityListenerHandler).GetMethod("OnShouldListenTo", BindingFlags.Static | BindingFlags.Public);
```

Comment:
nit: This line could be moved to the rest of the MethodInfo initializations in the `CreateActivityListenerInstance`